params/config.go

```diff
@@ -664,3 +617,69 @@ func (c *ChainConfig) GetFeeConfig() commontype.FeeConfig {
 func (c *ChainConfig) AllowedFeeRecipients() bool {
 	return c.AllowFeeRecipients
 }
+
+type ChainConfigWithUpgradesJSON struct {
+	ChainConfig
+	UpgradeConfig UpgradeConfig `json:"upgrades,omitempty"`
+}
+
+// MarshalJSON implements json.Marshaler. This is a workaround for the fact that
+// the embedded ChainConfig struct has a MarshalJSON method, which prevents
+// the default JSON marshalling from working for UpgradeConfig.
+// TODO: consider removing this method by allowing external tag for the embedded
+// ChainConfig struct.
+func (cu ChainConfigWithUpgradesJSON) MarshalJSON() ([]byte, error) {
+	// embed the ChainConfig struct into the response
+	chainConfigJSON, err := json.Marshal(cu.ChainConfig)
+	if err != nil {
+		return nil, err
+	}
+	if len(chainConfigJSON) > maxJSONLen {
+		return nil, errors.New("value too large")
+	}
+
+	type upgrades struct {
+		UpgradeConfig UpgradeConfig `json:"upgrades"`
+	}
+
+	upgradeJSON, err := json.Marshal(upgrades{cu.UpgradeConfig})
+	if err != nil {
+		return nil, err
+	}
+	if len(upgradeJSON) > maxJSONLen {
+		return nil, errors.New("value too large")
+	}
+
+	// merge the two JSON objects
+	mergedJSON := make([]byte, 0, len(chainConfigJSON)+len(upgradeJSON)+1)
+	mergedJSON = append(mergedJSON, chainConfigJSON[:len(chainConfigJSON)-1]...)
+	mergedJSON = append(mergedJSON, ',')
+	mergedJSON = append(mergedJSON, upgradeJSON[1:]...)
+	return mergedJSON, nil
+}
+
+func (cu *ChainConfigWithUpgradesJSON) UnmarshalJSON(input []byte) error {
+	var cc ChainConfig
+	if err := json.Unmarshal(input, &cc); err != nil {
+		return err
+	}
+
+	type upgrades struct {
+		UpgradeConfig UpgradeConfig `json:"upgrades"`
+	}
+
+	var u upgrades
+	if err := json.Unmarshal(input, &u); err != nil {
+		return err
+	}
+	cu.ChainConfig = cc
+	cu.UpgradeConfig = u.UpgradeConfig
+	return nil
+}
+
+func (c *ChainConfig) ToWithUpgradesJSON() *ChainConfigWithUpgradesJSON {
```

Comment:
can we add a comment about the purpose of this function?

Reply:
added